### PR TITLE
Add thread dumps to diagnostics

### PR DESCRIPTION
**Pull request by benclarkwood**
1. Adds an API to get a thread dump on-demand
2. Adds a thread dump to any Process.exit() call.
3. Grab thread dump of scheduler when test fails

### CODE REVIEW

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/framework/ProcessExit.java [hunk 1]
```diff
@@ -29,6 +33,8 @@ public static void exit(Code code) {
         String message = String.format("Process exiting immediately with code: %s[%d]", code, code.getValue());
         System.err.println(message);
         System.out.println(message);
+        System.err.println("Printing final thread state");
```
Comment (kvish) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/framework/ProcessExit.java]: nit: trailing dots to indicate action in progress to keep in line with other log statements

File: testing/sdk_diag.py [hunk 2]
```diff
@@ -162,6 +167,15 @@ def _dump_plans(item: pytest.Item, service_name: str):
             f.write('\n') # ... and a trailing newline
 
 
+def _dump_threads(item: pytest.Item, service_name: str):
+    threads = sdk_cmd.service_request('GET', 'v1/debug/threads')
+    out_path = _setup_artifact_path(item, 'threads_{}.json'.format(service_name.replace('/', '_')))
```
Comment (elezar) [inline testing/sdk_diag.py]: The threads endpoint doesn't output JSON though, correct? The filename seems to indicate this.

Also, would replacing / with __ not be more consistent?

Comment (benclarkwood) [reply testing/sdk_diag.py]: Good catch on the first point. I'm claiming "copying prior art" on the second one :).

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/http/endpoints/DebugResource.java [hunk 3]
```diff
@@ -24,11 +32,24 @@ public OfferOutcomeResource(OfferOutcomeTracker offerOutcomeTracker) {
...
             return ResponseUtils.htmlOkResponse(offerOutcomeTracker.toHtml());
         }
     }
+
+    /**
+     * Renders the current threads and thread state of the scheduler.
+     */
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    @Path("threads")
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/http/endpoints/DebugResource.java]: Could add this to the API doc: https://github.com/mesosphere/dcos-commons/blob/master/docs/reference/swagger-api/swagger-spec.yaml#L644

File: testing/sdk_diag.py [hunk 4]
```diff
@@ -162,6 +167,15 @@ def _dump_plans(item: pytest.Item, service_name: str):
             f.write('\n') # ... and a trailing newline
 
 
+def _dump_threads(item: pytest.Item, service_name: str):
+    threads = sdk_cmd.service_request('GET', 'v1/debug/threads')
```
Comment (elezar) [inline testing/sdk_diag.py]: There is a missing parameter here.

The `pylint -E testing` output:
```
************* Module sdk_diag
E:172,14: No value for argument 'service_path' in function call (no-value-for-parameter)
```
This should be:
```python
threads = sdk_cmd.service_request('GET', service_name, 'v1/debug/threads')
```

I have corrected this in #2532.

**Comment by porridge** [comment]
FTR, this does not work: shakedown insists on json by default, and server responds with a `406`.

**Pull request merged**